### PR TITLE
Do not destroy player entity completely on death

### PR DESCRIPTION
**Pull request by nihal111**
### Contains

An approach to not destroy the player entity completely upon death. This is done so as to allow for components attached to the player entity to persist after respawn.
Added death handling for character entities, that takes care of not destroying the player entity by consuming the BeforeDestroyEvent. The CharacterSystem captures the BeforeDestroy event and sends out a DeathEvent for all character entities regardless. There is just an additional check to consume the BeforeDestroy event (so as to save the DoDestroy from happening) in the case that the character entity is a Player (i.e. has a PlayerCharacterComponent). In that case, it also sends out a PlayerDeathEvent.

This adds the **OnPlayerRespawnedEvent, PlayerDeathEvent, AliveCharacterComponent, PlayerCharacterComponent**.

### Outstanding before merging

- [x] Respawn logic.
- [x] Fix things that break.
- [ ] Follow up: Add AliveCharacterComponent to all the character entities out there.

### Details

A character lives. It has the AliveComponent.
It dies.
BeforeDestroyEvent receives all entities having AliveComponent
+ If the entity *also has a PlayerCharacterComponent*, the event is consumed to prevent destruction. PlayerDeathEvent sent along with DeathEvent
+ Else just a DeathEvent sent
DeathScreen launches on DeathEvent and sends the respawnRequestEvent
respawnRequestEvent is received by the PlayerSystem
PlayerSystem then respawns player, sets the location for clientEntity to spawn location. Sends onPlayerRespawnedEvent

Components that need to be removed on death are removed by receiving the PlayerDeathEvent.
Components that need to be reset are done by receiving the onPlayerRespawnedEvent,

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/characters/CharacterSystem.java [hunk 1]
```diff
@@ -82,9 +83,13 @@
     @In
     private BlockEntityRegistry blockRegistry;
 
-    @ReceiveEvent(components = {CharacterComponent.class})
-    public void onDeath(DoDestroyEvent event, EntityRef entity) {
-        CharacterComponent character = entity.getComponent(CharacterComponent.class);
+    @ReceiveEvent
+    public void beforeDestroy(BeforeDestroyEvent event, EntityRef entity, CharacterComponent character) {
+        if (character.controller != EntityRef.NULL) {
```
Comment (flo) [inline engine/src/main/java/org/terasology/logic/characters/CharacterSystem.java]: This looks wrong. A destroyed entity will also behave like EntityRef.NULL but will have different id.

**Comment by flo** [comment]
I think this is the right direction but it is far from finished as there are the known issues you mentioned in the chat like movement during death screen. 

I think we should introduce an AliveComponent and use it to ensure that systems that are only intended for alive characters will only run when a character is alive.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/players/PlayerSystem.java [hunk 2]
```diff
@@ -88,6 +97,15 @@ public void update(float delta) {
                 i.remove();
             }
         }
+
+        i = clientsPreparingToRespawn.iterator();
+        while (i.hasNext()) {
+            SpawningClientInfo spawning = i.next();
+            if (worldProvider.isBlockRelevant(spawning.position)) {
+                respawnPlayer(spawning.clientEntity);
+                i.remove();
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/players/PlayerSystem.java]: Is this vulnerable to `ConcurrentModificationException` due to the iterator being modified while iterating?

File: engine/src/main/java/org/terasology/logic/players/PlayerCharacterComponent.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package org.terasology.logic.players;
+
+import org.terasology.entitySystem.Component;
+
+/**
+ * This is only attached to the player entities. <br/>
+ * Used to differentiate between player characters and other character entities.
+ */
+public class PlayerCharacterComponent implements Component {
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/players/PlayerCharacterComponent.java]: Do we have any existing components that are unique to a PC?

File: engine/src/main/java/org/terasology/logic/players/PlayerSystem.java [hunk 4]
```diff
@@ -187,14 +205,36 @@ public void onRespawnRequest(RespawnRequestEvent event, EntityRef entity) {
...
+        ClientComponent client = clientEntity.getComponent(ClientComponent.class);
+        EntityRef playerCharacter = client.character;
+        LocationComponent location = clientEntity.getComponent(LocationComponent.class);
+        PlayerFactory playerFactory = new PlayerFactory(entityManager, worldProvider);
+        Vector3f spawnPosition = playerFactory.findSpawnPositionFromLocationComponent(location);
+        location.setWorldPosition(spawnPosition);
+        clientEntity.saveComponent(location);
+
+        AliveCharacterComponent aliveCharacterComponent = new AliveCharacterComponent();
+        playerCharacter.addComponent(aliveCharacterComponent);
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/players/PlayerSystem.java]: Could just short-cut these two lines and `addComponent(new AliveCharacterComponent():`

File: engine/src/main/resources/assets/prefabs/player/player.prefab [hunk 5]
```diff
@@ -1,52 +1,83 @@
...
-        "diveVolume": 0.5,
-        "landingVolume": 0.25
-    },
-    "CharacterHeldItem": {
-    },
-    "Breather": {},
-    "GazeMountPoint": {
-        "translate": [0, 0.7, 0]
-    }
+  "persisted": true,
```
Comment (Cervator) [inline engine/src/main/resources/assets/prefabs/player/player.prefab]: Looks like this file went from 4 space indents to 2 space indents. I think we've got some inconsistency in json files.

File: engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java [hunk 6]
```diff
@@ -28,6 +28,9 @@
 import java.util.List;
 
 /**
+ * This component is attached to all character entities. It governs movement and stores
+ * associated paramenters. <br/>
+ * The AliveComponent should necessarily be attached to the character entity for the movement systems to work.
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java]: Maybe outdated component name in doc

File: engine/src/main/java/org/terasology/logic/players/PlayerCharacterComponent.java (hunk 3 above)
Comment (nihal111) [inline engine/src/main/java/org/terasology/logic/players/PlayerCharacterComponent.java]: I guess not. All current checks look for a `characterComponent.controller` to exist to determine if the character entity is a player (controlled by a client) or not.
According to @flo this should change as an AI/NPC can be connected to a client too.

File: engine/src/main/resources/assets/prefabs/player/player.prefab (hunk 5 above)
Comment (nihal111) [inline engine/src/main/resources/assets/prefabs/player/player.prefab]: 2 space is our standard for json. Some ancient files haven't be changed.

File: engine/src/main/java/org/terasology/logic/players/PlayerSystem.java (hunk 2 above)
Comment (nihal111) [inline engine/src/main/java/org/terasology/logic/players/PlayerSystem.java]: No, removing using the iterator is the safe way to do it. 
https://stackoverflow.com/questions/18448671/how-to-avoid-concurrentmodificationexception-while-removing-elements-from-arr

File: engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java (hunk 6 above)
Comment (nihal111) [inline engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java]: Not sure what you mean.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java (hunk 6 above)
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/characters/CharacterMovementComponent.java]: `AliveComponent` vs `CharacterAliveComponent` - was renamed but remains old name in doc :-)

**Comment by flo** [comment]
Tests out good with what I have tried.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by flo** [comment]
Finally we can keep the inventory after respawn :)

**Comment by nihal111** [comment]
A lot more than inventory :wink:

**Pull request merged**